### PR TITLE
Implement database compression

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
|Related issue|
|---|
|#20163|

## Objective

This pull request addresses the need to improve the storage efficiency of our database system by integrating the `BZIP2 `compression algorithm into the RocksDB engine. The motivation for employing `BZIP2` compression lies in its ability to achieve high compression ratios while maintaining a balance between compression speed and decompression performance.

## Description
- Implemented the wrapper for database compression.
- Added compression for candidates and CVE5 databases.
- Included benchmark tests to verify the accuracy of the validation process.

## Quality Assurance
A benchmark between the `LZ4HC`, `LZ4`, and `BZIP2` algorithms was made, using a database of 100.000 elements. 

The original database size was `~3.4GB` for CVEs, and the final result is:

| Algorithm | Size after compression | Time to compress | Size improvement percent |
|-----------|------|------|------------------|
| LZ4HC | ~565MB| 44.75s | ~83.4%|
| BZIP2 | ~476MB| 522.57s | ~86%|
| LZ4 | ~668MB| 29.34s | ~80.35%|

We can see that `BZIP2` has better performance in reducing the size of the database but, the difference between this algorithm and the `LZ4HC` in size is minimal, but in time is `x4`faster than `LZ4HC`.

However, we already have the `BZIP2` library in our repository, and in order to avoid adding an extra dependence we use the `BZIP2`.

No new test has been included.

Regarding memory, no new memory allocations were introduced in this PR using `raw_pointers`.

<details><summary>ScanBuild report</summary>

Command:

```bash
scan-build-15 -enable-checker alpha.core.CastSize -enable-checker alpha.core.CastToStruct -enable-checker alpha.core.Conversion -enable-checker alpha.core.IdenticalExpr -enable-checker alpha.core.SizeofPtr -enable-checker alpha.core.TestAfterDivZero --exclude external make TARGET=server DEBUG=1 -j$(nproc)
```

```bash

Done building server

scan-build: No bugs found.

```

</details>

<details><summary>Clang-Format</summary>

```bash
---------------------------------
PASSED: Clang-format check passed
---------------------------------
```
</details>

## Exploratory tests
Not apply

### Check summary
- [x] Style and quality of SCA.
- [x] Documentation clear.
- [x] Functionality optimizated.

### CODE REVIEW

File: src/shared_modules/utils/rocksDBWrapper.hpp [hunk 1]
```diff
@@ -31,6 +31,7 @@ namespace Utils
         {
             rocksdb::Options options;
             options.create_if_missing = true;
+            options.compression = rocksdb::kLZ4HCCompression;
```
Comment (pereyra-m) [inline src/shared_modules/utils/rocksDBWrapper.hpp:34]: I've found that `LZ4` is always recommended over the default Snappy
https://github.com/facebook/rocksdb/wiki/Compression

But are we sure that `LZ4HC` is better than `LZ4`?
Here we have some benchmarks 

http://www.lmdb.tech/bench/inmem/compress/RocksDB/

File: src/shared_modules/utils/rocksDBWrapper.hpp [hunk 2]
```diff
@@ -238,6 +239,26 @@ namespace Utils
...
+         *
+         * @see rocksdb::CompactRangeOptions
+         */
+        void compactDatabase()
+        {
+            // Create compact range options with default settings
+            rocksdb::CompactRangeOptions compactOptions;
+
+            // Perform compaction for the entire key range
+            m_db->CompactRange(compactOptions, nullptr, nullptr);
```
Comment (pereyra-m) [inline src/shared_modules/utils/rocksDBWrapper.hpp:259]: This seems to be an advanced feature that we should fully understand before including it
https://github.com/facebook/rocksdb/wiki/Manual-Compaction

It doesn't seem to be related to the compression, it's more like  a "vacuum" .
But if we are going to use it anyway, we shouldn't run it after every message processed. It could be only at startup for example.

**Review by pereyra-m — CHANGES_REQUESTED** [review]
GJ!

File: src/shared_modules/utils/rocksDBWrapper.hpp (hunk 1 above)
Comment (GabrielEValenzuela) [reply src/shared_modules/utils/rocksDBWrapper.hpp:34]: I made the test with the three algorithms and wrote in the PR summary. 
I agree with the web page, `zlib` is the best option but is very slow. `zl4hc` has good performance and time rather than `lz4`

File: src/shared_modules/utils/rocksDBWrapper.hpp (hunk 2 above)
Comment (GabrielEValenzuela) [reply src/shared_modules/utils/rocksDBWrapper.hpp:259]: I believe that the concept of compacting the CVE database is well-defined. This particular database is not likely to be used and aligns with the second point on the list: "Force the data to go through the compaction filter in order to consolidate it."

Also, we call the compaction once all messages have been processed.

File: src/shared_modules/utils/rocksDBWrapper.hpp (hunk 1 above)
Comment (pereyra-m) [reply src/shared_modules/utils/rocksDBWrapper.hpp:34]: Great, thank you!

File: src/shared_modules/utils/rocksDBWrapper.hpp (hunk 2 above)
Comment (pereyra-m) [reply src/shared_modules/utils/rocksDBWrapper.hpp:259]: Ok, thank you.

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp [hunk 3]
```diff
@@ -178,7 +178,14 @@ class TDatabaseFeedManager final : public Observer<nlohmann::json&>
                 };
                 try
                 {
-                    TMessageProcessor::processMessage(message, orchestrationLambda);
+                    std::cout << "Messages processed successfully." << std::endl;
```
Comment (pereyra-m) [inline src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp:181]: You've deleted the line
```
TMessageProcessor::processMessage(message, orchestrationLambda);
```

**Review by pereyra-m — CHANGES_REQUESTED** [review]

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp (hunk 3 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp]: Done!

**Review by pereyra-m — APPROVED** [review]
LGTM!

**Comment by Dwordcito** [comment]
@GabrielEValenzuela please add benchmark metrics.

**Comment by GabrielEValenzuela** [comment]
@Dwordcito 

Here are the results of the benchmark

```text
 CPU: Intel i5-10500H (12) @ 4.500GHz 
GPU: NVIDIA GeForce RTX 3050 Mobile 
Memory: 15903MiB 
OS: Manjaro Linux x86_64 
```

|      | CVEs  | Description | Remediation | Translastion | Feeds  |
|------|-------|-------------|-------------|--------------|--------|
| Time | 430s  | 1s          | 52ms        | 18us         | 15s    |
| Size | 477MB | 134MB       | 5MB         | 37kB         | ~1.7GB |

Regarding the CPU use, only one thread works at 100%, not all the time.

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp [hunk 4]
```diff
@@ -180,6 +180,20 @@ class TDatabaseFeedManager final : public Observer<nlohmann::json&>
                     logInfo(WM_VULNSCAN_LOGTAG, "Processing message");
                     TMessageProcessor::processMessage(message, orchestrationLambda);
                     logInfo(WM_VULNSCAN_LOGTAG, "Message processed");
+
+                    // Compact the database
+                    logInfo(WM_VULNSCAN_LOGTAG, "Compacting CVE5 database");
+                    m_cvesDatabase->compactDatabase();
```
Comment (pereyra-m) [inline src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp:186]: Please, use the method with bzip2 compression

**Review by pereyra-m — CHANGES_REQUESTED** [review]
Gj!

### Test

The suggested change was implemented locally and a single document was processed with and without the new compression algorithm (the branch was rebased to include the main branch fixes)

<details><summary>Details</summary>
<p>


```
root@ubuntu-jammy:/home/vagrant/wazuh/src/build/wazuh_modules/vulnerability_scanner/testtool/databaseFeedManager# du -hs queue_no_compactation/vd/*
36M     queue_no_compactation/vd/candidates
42M     queue_no_compactation/vd/cves
664K    queue_no_compactation/vd/descriptions
200K    queue_no_compactation/vd/remediations
48K     queue_no_compactation/vd/translations
root@ubuntu-jammy:/home/vagrant/wazuh/src/build/wazuh_modules/vulnerability_scanner/testtool/databaseFeedManager# du -hs queue_without_bzip2/vd/*
19M     queue_without_bzip2/vd/candidates
42M     queue_without_bzip2/vd/cves
656K    queue_without_bzip2/vd/descriptions
76K     queue_without_bzip2/vd/remediations
48K     queue_without_bzip2/vd/translations
root@ubuntu-jammy:/home/vagrant/wazuh/src/build/wazuh_modules/vulnerability_scanner/testtool/databaseFeedManager# du -hs queue_without_bzip2/vd/*
19M     queue_without_bzip2/vd/candidates
42M     queue_without_bzip2/vd/cves
656K    queue_without_bzip2/vd/descriptions
76K     queue_without_bzip2/vd/remediations
48K     queue_without_bzip2/vd/translations
```

</p>
</details> 

@GabrielEValenzuela , have you noticed an improvement with the bzip2 compression?
I haven't in this test

### Test update

I've tested again with more files and now I notice the difference

<details><summary>Details</summary>
<p>


```
root@ubuntu-jammy:/home/vagrant/wazuh/src/build/wazuh_modules/vulnerability_scanner/testtool/databaseFeedManager# du -hs queue_without_compression/vd/*
111M    queue_without_compression/vd/candidates
152M    queue_without_compression/vd/cves
6.1M    queue_without_compression/vd/descriptions
4.2M    queue_without_compression/vd/remediations
48K     queue_without_compression/vd/translations
root@ubuntu-jammy:/home/vagrant/wazuh/src/build/wazuh_modules/vulnerability_scanner/testtool/databaseFeedManager# du -hs queue_withou_bzip2/vd/*
79M     queue_withou_bzip2/vd/candidates
152M    queue_withou_bzip2/vd/cves
6.0M    queue_withou_bzip2/vd/descriptions
1012K   queue_withou_bzip2/vd/remediations
48K     queue_withou_bzip2/vd/translations
root@ubuntu-jammy:/home/vagrant/wazuh/src/build/wazuh_modules/vulnerability_scanner/testtool/databaseFeedManager# du -hs queue_with_bzip2/vd/*
79M     queue_with_bzip2/vd/candidates
24M     queue_with_bzip2/vd/cves
6.0M    queue_with_bzip2/vd/descriptions
1008K   queue_with_bzip2/vd/remediations
48K     queue_with_bzip2/vd/translations
```


</p>
</details>

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp (hunk 4 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp:186]: Yes sorry! I forgot it!

**Review by pereyra-m — APPROVED** [review]
